contracts/src/v2/Calls.sol

```diff
@@ -0,0 +1,145 @@
+import {AgentExecutor} from "../AgentExecutor.sol";
+import {Agent} from "../Agent.sol";
+import {Call} from "../utils/Call.sol";
+import {Token} from "../Token.sol";
+import {Upgrade} from "../Upgrade.sol";
+import {Functions} from "../Functions.sol";
+import {Constants} from "../Constants.sol";
+
+import {Ticket, TransferKind, OperatingMode} from "./Types.sol";
+
+import {UD60x18, ud60x18, convert} from "prb/math/src/UD60x18.sol";
+
+/// @title Library for implementing Ethereum->Polkadot ERC20 transfers.
+library CallsV2 {
+    using Address for address;
+    using SafeTokenTransfer for IERC20;
+    using SafeNativeTransfer for address payable;
+
+    error InvalidProof();
+    error InvalidNonce();
+    error NotEnoughGas();
+    error FeePaymentToLow();
+    error Unauthorized();
+    error Disabled();
+    error AgentAlreadyCreated();
+    error AgentDoesNotExist();
+    error ChannelAlreadyCreated();
+    error ChannelDoesNotExist();
+    error InvalidChannelUpdate();
+    error AgentExecutionFailed(bytes returndata);
+    error InvalidAgentExecutionPayload();
+    error InvalidConstructorParams();
+    error AlreadyInitialized();
+    error TokenNotRegistered();
+
+    error InvalidAsset();
+
+    address public constant WETH_ADDRESS = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
+
+    // Send an XCM with assets to Polkadot Asset Hub
+    //
+    // Params:
+    //   * `xcm` (bytes): SCALE-encoded XCM message
+    //   * `assets` (bytes[]): Array of asset transfer instructions
+    //
+    // The specified assets will be locked/burned locally, and their foreign equivalents will
+    // be minted/unlocked on Polkadot Asset Hub.
+    //
+    // Supported asset instructions:
+    // * ERC20: abi.encode(0, tokenAddress, value)
+    //
+    function sendMessage(bytes calldata xcm, bytes[] calldata assets) external {
+        bytes[] memory encodedAssets = new bytes[](assets.length);
+        for (uint256 i = 0; i < assets.length; i++) {
+            encodedAssets[i] = _handleAsset(assets[i]);
+        }
+
+        Ticket memory ticket =
+            Ticket({origin: msg.sender, assets: encodedAssets, xcm: xcm});
+        _submitOutbound(ticket);
+    }
+
+    /*
+    * Internal functions
+    */
+
+    // Submit an outbound message to Polkadot, after taking fees
+    function _submitOutbound(Ticket memory ticket) internal {
+        CoreStorage.Layout storage $ = CoreStorage.layout();
+
+        // Ensure outbound messaging is allowed
+        _ensureOutboundMessagingEnabled();
+
+        // Wrap provided ether and transfer to AssetHub agent
+        address assetHubAgent = Functions.ensureAgent(Constants.ASSET_HUB_AGENT_ID);
+        WETH9(payable(WETH_ADDRESS)).deposit{value: msg.value}();
+        IERC20(WETH_ADDRESS).safeTransfer(assetHubAgent, msg.value);
+
+        $.outboundNonce = $.outboundNonce + 1;
+
+        bytes memory payload =
+            SubstrateTypes.encodePayloadV2(ticket.origin, ticket.assets, ticket.xcm);
```

Comment:
Cool, A lot more simple now!

With `ticket.xcm` does that mean for V2 there is no need to construct xcm on chain any more?

UI will build the xcm and it's the responsibility of relayer to validate the xcm and check the relay is profitable or not.

On BH we just decode the xcm with
```
VersionedXcm::<()>::decode_with_depth_limit(
    MAX_XCM_DECODE_DEPTH,
    &mut data,
)
``` 
and forward it to AH transparently.

Reply:
Yeah, exactly.

Though for registration of native tokens, we need to control the XCM and so we need to build it on chain. See more details in https://github.com/Snowfork/snowbridge/pull/1300/commits/149c3e425bbcaf023a9564977921bd355cf129fd.

Reply:
> Though for registration of native tokens, control the XCM and so we need to build it on chain

Still not sure fully understand why `ticket.assets` is required here, why not just build the xcm off chain?

@vgeddes I've removed the `MessageConverter` completely in https://github.com/yrong/polkadot-sdk/pull/3 which I'd assume is not in use any more, correct me if I'm wrong.

Reply:
> Still not sure fully understand why ticket.assets is required here, why not just build the xcm off chain?

When the message gets to BH, we convert the `payload.assets` to instructions like ReserveAssetDeposited,etc, and then append the instructions in `payload.xcm`.

Or in other words, BH builds final XCM using:
```
finalXCM = concat(
assets_to_xcm(payload.asssets),
UniversalOrigin(Ethereum)
DescendOrigin(payload.origin)
payload.xcm
)
```

We can't trust the user to build the entire xcm offchain as they could mint any assets they want. Rather, they can only supply a fragment of the xcm.

Reply:
> We can't trust the user to build the entire xcm offchain as they could mint any assets they want.

Instead of convert the assets to xcm, can we just do some basic check on BH to verify the xcm is constructed based upon the assets correctly(e.g.  `ensure!(xcm contains ReserveAssetDeposited(asset)|WithdrawAsset(asset)`)?

So the point here is validate rather than construct.

Reply:
Btw: for transfer use case I'm not sure `DescendOrigin(userOrigin)` make sense. For PNA we need to burn the asset locked in Ethereum sovereign on AH. That's why we need to change origin to `UniversalOrigin(GlobalConsensus(network))` before `WithdrawAsset` [here](https://github.com/paritytech/polkadot-sdk/blob/21b3a46b11c15ff3915557201657322f42527a2b/bridges/snowbridge/primitives/router/src/inbound/mod.rs#L440-L441).

Reply:
> it is too late because we have already changed the state on the Ethereum side

Just curious what will happen in this case. Is this something acceptable assuming we will implement error handing mechanism any way? The errored xcm will get trapped on BH and we allow end user to repair it? 

Something like we allow user to add fee for message pending from Substrate to Ethereum direction, assuming there is no relayer who will pick it up if it's not profitable.

Reply:
> Just curious what will happen in this case. Is this something acceptable assuming we will implement error handing mechanism any way?

I think if the message is queued and if the message is profitable, than once submitted to BH we should not fail for any reason (such as validating the xcm for example) and we should try to always push the assets to AH where they can get trapped if the user passes invalid xcm. And we should trust that we constrain the origin to stop users from doing things that they cannot within the xcm, instead of validating.